# Evaluation and Tuning

This notebook is to document evaluation trials analysis and log changes after each tuning approach

## Datasets:

- **evaluation.csv**: 100 Statistics theory and applications questions with ground truths and ground contexts


## Evaluation approach:

- **LLM-as-a-judge**: Using LLMs to justify the answer and give evaluation metrics. The metrics generated by this method are: correctness and faithfulness

- **Framework**: **LangSmith** platform to monitor and record experiments.

## Evaluation metrics used:

### Correctness

- **What it measures:** Internal logical consistency.

- **The Definition:** Does the model's reasoning logically support the final answer it gave to the user's question?

- **How it works:** It looks only at the Question, the Answer, and the Reasoning. It does not check if the facts are true in the real world; it only checks if the model's thought process makes logical sense from point A to point B.

### Faithfulness

- **What it measures:** Groundedness and absence of hallucinations.

- **The Definition:** Is every single claim in the generated answer entirely backed up by the retrieved documents?

- **How it works:** It looks at the Answer and the Retrieved Context. If the model brings in outside knowledge that isn't in your text chunks—even if that outside knowledge is technically true—it fails the faithfulness check.

### Notes

- All LLMs in used were hosted locally via Ollama. Hence, I can evaluate the system expansively without worrying about cost. However, the latency is high since my machine is not as powerful as corporate systems :(

## Experiment: glossy-root-69 (Baseline)

### Dataset: 'evaluations.csv'
### RAG schema: Ingestion -> Chunking (size = 200, overlap =  100) -> BM_25 (0.7, k=3) + ChromaDB Retriever (0.3, k=3) -> Google Gemma 3 (Ollama)
### Metris: Correctness (DeepSeek-R1)

In [1]:
from langsmith import Client
from dotenv import load_dotenv
from src.utils import beautify_langsmith_stats

load_dotenv()

client = Client()
exp = client.get_experiment_results(name='glossy-root-69')
print(beautify_langsmith_stats(exp))

### Evaluation Summary
* **Feedback: valid_reasoning**
        - **Average Score:** 76.3% (0.763)
        - **Standard Deviation:** 0.425
        - **Evaluated Runs:** 452 (Errors: 0)
* **Run Performance & Telemetry**
        - **Total Runs:** 453
        - **Latency:** 1.27s (Median) | 3.23s (99th Percentile)
        - **Token Usage:** 327,427 Total (269,282 prompt / 58,145 completion)
        - **Last Run:** March 14, 2026 at 06:13 AM


## Experiment: dependable-loss-22

### Dataset: 'evaluations.csv'
### RAG schema: Ingestion -> Chunking (size = 200, overlap =  100) -> BM_25 (0.7, k=3) + ChromaDB Retriever (0.3, k=3) -> Google Gemma 3 (Ollama)
### Metris: Correctness (DeepSeek-R1), Faithfulness (Llama3.1)

In [1]:
from langsmith import Client
from dotenv import load_dotenv
from src.utils import beautify_langsmith_stats

load_dotenv()

client = Client()
exp = client.get_experiment_results(name='dependable-loss-22')
print(beautify_langsmith_stats(exp))

### Evaluation Summary

* **Feedback: check_faithfulness**
        - **Average Score:** 92.6% (0.926)
        - **Standard Deviation:** 0.184
        - **Evaluated Runs:** 309 (Errors: 0)
* **Feedback: valid_reasoning**
        - **Average Score:** 83.2% (0.832)
        - **Standard Deviation:** 0.374
        - **Evaluated Runs:** 309 (Errors: 0)
* **Run Performance & Telemetry**
        - **Total Runs:** 310
        - **Latency:** 3.57s (Median) | 5.18s (99th Percentile)
        - **Token Usage:** 223,642 Total (184,506 prompt / 39,136 completion)
        - **Last Run:** March 14, 2026 at 08:27 AM


## Experiment: new-stone-19

### Dataset: 'evaluations.csv'
### RAG schema: Ingestion -> Chunking (size = 250, overlap =  100) -> BM_25 (0.7, k=3) + ChromaDB Retriever (0.3, k=3) -> Google Gemma 3 (Ollama)
### Metris: Correctness (DeepSeek-R1), Faithfulness (Llama3.1)

In [2]:
from langsmith import Client
from dotenv import load_dotenv
from src.utils import beautify_langsmith_stats

load_dotenv()

client = Client()
exp = client.get_experiment_results(name='new-stone-19')
print(beautify_langsmith_stats(exp))

### Evaluation Summary
* **Feedback: check_faithfulness**
        - **Average Score:** 87.8% (0.878)
        - **Standard Deviation:** 0.223
        - **Evaluated Runs:** 144 (Errors: 0)
* **Feedback: valid_reasoning**
        - **Average Score:** 79.9% (0.799)
        - **Standard Deviation:** 0.401
        - **Evaluated Runs:** 144 (Errors: 0)

* **Run Performance & Telemetry**
        - **Total Runs:** 145
        - **Latency:** 3.47s (Median) | 5.17s (99th Percentile)
        - **Token Usage:** 117,763 Total (98,834 prompt / 18,929 completion)
        - **Last Run:** March 14, 2026 at 08:32 PM


# Conclusion:

- Using 2 LLMs to evaluate increased the latency of each evaluation. However, the system performed well with the setup (chunk size: 200, chunk overlap: 100).
- Both correctness and faithfulness decreased after increasing chunk size and chunk overlap


## Experiment: essential-truck-74

### Dataset: 'evaluations.csv'
### RAG schema: Ingestion -> Chunking (size = 200, overlap =  100) -> BM_25 (0.5, k=10) + ChromaDB Retriever (0.5, k=10) -> Google Gemma 3 (Ollama)
### Metris: Correctness (DeepSeek-R1), Faithfulness (Llama3.1)
### Pipeline modification: Used Pydantic to enforce LLM answer to use citations all the times.

In [1]:
from langsmith import Client
from dotenv import load_dotenv
from src.utils import beautify_langsmith_stats

load_dotenv()

client = Client()
exp = client.get_experiment_results(name='essential-truck-74')
print(beautify_langsmith_stats(exp))

### Evaluation Summary
* **Feedback: check_faithfulness**
        - **Average Score:** 84.8% (0.848)
        - **Standard Deviation:** 0.220
        - **Evaluated Runs:** 108 (Errors: 0)
* **Feedback: valid_reasoning**
        - **Average Score:** 75.9% (0.759)
        - **Standard Deviation:** 0.428
        - **Evaluated Runs:** 108 (Errors: 0)

* **Run Performance & Telemetry**
        - **Total Runs:** 109
        - **Latency:** 4.69s (Median) | 366.09s (99th Percentile)
        - **Token Usage:** 424,180 Total (156,369 prompt / 267,811 completion)
        - **Last Run:** March 15, 2026 at 03:17 AM


### Verdict:

I was hoping to get a decent result as the actual question hand tests went very well. After enforcing by Pydanctic, Gemma 3 has been producing consistently formatted results (with citations included). However, when Gemma produces unexpected citations or forgot to include, Pydantic could not understand and behave unexpectedly, too.

With the previous setup, although the faithfulness and correctness were high, the model usually gives answers with no citations. So, I suspect the model was using its trained knowledge to produce answers. When checking back the responses of this setup manually, the entries with low correctness were actually decent. As a result, I examined a case where the judge labeled 0 correctness

- Inputs:
    - **Question:** Why is accuracy misleading for imbalanced data?
- Outputs:
    - **Answer:** According to Devore (2007), \"value is sometimes deceptive because it can be greatly inflated by using a large number of predictors relative to the sample size\" (Page: 582). This means that when dealing with imbalanced data, a simple accuracy measure can be misleading because the model might be overfitting to the minority class, leading to an artificially high accuracy score that doesn't reflect the true predictive performance (Devore, 2007, Page: 488).
    - **Context (from model's answer):** A model can just predict the majority class and score high accuracy while missing the minority class.
    - **Ground truth:** It ignores performance on the minority class.

Even though it sounds correct, the concepts were mixed showing the LLM was hallucinated. This might be because too many documents were fed to it from RAG.

# Current Optimized setup

#### RAG schema: Ingestion -> Chunking (size = 200, overlap =  100) -> BM_25 (0.7, k=3) + ChromaDB Retriever (0.3, k=3) -> Google Gemma 3 (Ollama)

# Sanity Check for Retrievers

In [9]:
from langchain_ollama import ChatOllama
from src.vector_db_utils import Database
import pandas as pd
from pathlib import Path
from tqdm import tqdm
import ollama


cur_path = Path(__name__).parent.resolve()

db = Database()
ret = db.get_retriever()
df = pd.read_csv(f'{cur_path.parent / 'evaluation.csv'}')
llm = ChatOllama(model='gemma3', temperature=0)

def format_string(docs):
    context = []
    for d in docs:
        context.append(d.page_content)
    return "\n\n--\n\n".join(context)

system_instructions = """You are an expert evaluator grading a search engine's retrieval quality.
Analyze the provided context and the ground truth.

Output STRICTLY a single integer: the number of claims in contexts can be used to answer the question
Do not output any other text, formatting, or reasoning."""

human_template = """Question: {question}
Context: {context}
Ground Truth: {ground_truth}"""

# 2. Build the chain.
# We use itemgetter to pluck the right variables from the input dictionary.
# Make sure your retriever output is formatted into a string!


scores = 0
for index, row in tqdm(df.iterrows(), total=len(df)):

    # 1. Package the row data into the dictionary the chain expects
    context = format_string(ret.invoke(row['question']))
    msg = f"Question: {row['question']}\nContext: {context}\nGround Truth: {row['ground_truth']}"

    response = ollama.chat(
        model="gemma3",
        messages=[
            {"role": "system", "content": system_instructions},
            {"role": "user", "content": msg}
        ],

        options={"temperature": 0.0}
    )

    scores += int(response.message.content) / 10
    # 2. Run the evaluation
    # print(response.content)
print(f"Retrieval Score: {(scores/100):2f}%")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 12115.81it/s]
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
100%|██████████| 100/100 [00:44<00:00,  2.25it/s]

Retrieval Score: 0.368000%


### Confirmed retrieval system is not performing okay as retrieval precision is only 36.8%. However, when plugging into LLM, the model is still able to achieve considerable scores. Hence, current usage is okay but requires more metadata, better cleaning, and different retrieval methods. In fact, many of the chunks could not capture the whole context.

# Example outputs

In [11]:
from src.pipeline import pipeline_enforced
from langchain_ollama import ChatOllama
model = ChatOllama(model='gemma3', temperature=0)

chain = pipeline_enforced(model=model)
chain.invoke("An e-commerce site wants to test a new checkout flow, but their daily traffic is very low. They want to peek at the results every day to stop the test early if there's a winner. What statistical framework allows for this continuous monitoring without inflating the Type I error rate?")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 11443.45it/s]
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


RAGResponse(answer='The statistical framework that allows for continuous monitoring of a test like this without inflating the Type I error rate is a Bonferroni correction (Devore, 2023). This method adjusts the significance level (alpha) for multiple comparisons. Since the e-commerce site is observing the results daily and potentially stopping the test early, they are conducting multiple comparisons. The Bonferroni correction divides the desired alpha level (e.g., 0.05) by the number of comparisons to control the overall Type I error rate (Devore, 2023).', citations=[Citation(author='Devore, Jay L.', title='Probability and Statistics for Engineering and the Sciences, 8th ed.', page='[330]', year=2023)])

In [12]:
chain.invoke("A SaaS company tracks customer churn. Some customers cancel after 3 months, but many are still active at the end of the 2-year observation period. What specific type of data do these active customers represent in survival analysis?")

RAGResponse(answer='These active customers represent the *censored* data in survival analysis. In survival analysis, censoring refers to situations where a customer’s event of interest (churn) hasn’t been observed by the end of the study period (2 years) (Devore, 2023). Because these customers remained active, their event (churn) was not observed, thus they are considered censored (Unknown, 2023). “Censoring is a critical concept in survival analysis, as it represents data points where the event of interest has not yet occurred” (Unknown, 2023, p. 346).\n\nCitation:\nDevore, J. L. (2023). *Probability and Statistics for Engineering and the Sciences, 8th ed.* Macmillan.\nUnknown. (2023). *Testing Statistical Assumptions*.\n', citations=[])